File: Programmers/kakao/순위 검색.py
def solution(info, query):
    answer = []
    # 0:개발언어, 1:직군, 2:경력, 3:소울푸드, 4:점수
    score = dict()

    for information in info:
        temp = information.split(' ')
        if int(temp[4]) in score:
            score[int(temp[4])].append(temp[:4])
        else:
            score[int(temp[4])] = [temp[:4]]
    score = sorted(score.items())

    for commands in query:
        command = ''.join(commands.split('and')).split()
        left = 0
        right = len(score) - 1
        target = int(command[-1])
        while left <= right:
            mid = (left+right) // 2
            if target > score[mid][0]:
                left = mid + 1
            elif target <= score[mid][0]:
                right = mid - 1
        count = 0

        for i in range(left, len(score)):
            for item in score[i][1]:
                if (item[0] == command[0] or command[0] == '-') and (item[1] == command[1] or command[1] == '-') and (item[2] == command[2] or command[2] == '-') and (item[3] == command[3] or command[3] == '-'):
                    count += 1
        answer.append(count)

    return answer


solution(["java backend junior pizza 150", "python frontend senior chicken 210", "python frontend senior chicken 150",
          "cpp backend senior pizza 260", "java backend junior chicken 80", "python backend senior chicken 50"],
         ["java and backend and junior and pizza 100", "python and frontend and senior and chicken 200",
          "cpp and - and senior and pizza 250", "- and backend and senior and - 150", "- and - and - and chicken 100",
          "- and - and - and - 150"])

File: Java/src/com/company/level2/순위검색.java
package com.company.level2;

import java.util.*;

public class 순위검색 {
    public int[] solution(String[] info, String[] query) {
        int[] answer = new int[query.length];
        Map<String, List<Integer>> map = new HashMap<>();
        String[] lan = {"cpp", "java", "python", "-"};
        String[] skill = {"backend", "frontend", "-"};
        String[] career = {"junior", "senior", "-"};
        String[] food = {"chicken", "pizza", "-"};

        for(int i = 0; i < 4; i++){
            for(int j = 0; j < 3; j++){
                for(int k = 0; k < 3; k++){
                    for(int r = 0; r < 3;r ++){
                        map.put(lan[i]+skill[j]+career[k]+food[r], new ArrayList<>());
                    }
                }
            }
        }

        for(String i : info){
            Stack<String> stack = new Stack<>();
            List<String> list = new ArrayList<>();
            String[] ia = i.split(" ");
            dfs(0, ia, stack, list);

            for(int j = 0; j < list.size(); j++){
                map.get(list.get(j)).add(Integer.parseInt(ia[4]));
            }
        }

        for (String s : map.keySet()) {
            Collections.sort(map.get(s));
        }



        int i = 0;
        for (String q : query){
            String[] qa = q.split(" ");
            String s = qa[0] + qa[2] + qa[4] + qa[6];
            int point = Integer.parseInt(qa[7]);
            List<Integer> list = map.get(s);

            int left = 0;
            int right = list.size()-1;
            int start = 0;
            while(left <= right){
                int mid = (left + right) / 2;

                if(point <= list.get(mid))
                    right = mid-1;
                else
                    left = mid + 1;

            }


            answer[i] = list.size() - left;

            i++;
        }

        return answer;
    }

    public void dfs(int depth, String[] sa, Stack<String> stack, List<String> results){
        if(depth == sa.length-1){
            String result = "";
            for(String s : stack){
                result += s;
            }
            results.add(result);
            return;
        }

        String[] temp = {sa[depth], "-"};

        for(int i = 0; i < 2; i++){
            stack.add(temp[i]);
            dfs(depth+1, sa, stack, results);
            stack.pop();
        }

    }

}
